lib/mail/encoders/eight_bit.ex

```diff
@@ -0,0 +1,47 @@
+defmodule Mail.Encoders.EightBit do
+  @moduledoc """
+  Encodes/decodes 8-bit strings according to RFC 2045.
+
+  See the following link for reference:
+  - <https://tools.ietf.org/html/rfc2045#section-2.8>
+  """
+
+  @new_line "\r\n"
+  @wrap_length 998
+
+  @doc """
+  Encodes a string into an 8-bit encoded string.
+
+  Raises if any character is <NUL> "\0".
+  """
+  def encode(string), do: do_encode(string, "", 0)
+  defp do_encode(<<>>, acc, _line_length), do: acc
+  defp do_encode(<<head, tail::binary>>, acc, line_length) do
+    {encoding, line_length} = emit_char(head, line_length)
+    do_encode(tail, acc <> encoding, line_length)
+  end
+
+  defp emit_char(?\0, _line_length) do
+    raise ArgumentError, message: "illegal NUL character"
+  end
+  defp emit_char(char, line_length) do
```

Comment:
Funny, I actually implemented it that way initially. But then I thought it would be easier to understand this way. I'm new to Elixir so I'm still feeling out when it's best to use guard clauses.